Extend client tracking tests

nitaicaro opened the pull request:
Test support for the new map, null and push message types. Map objects are parsed as a list of lists of key value pairs.
for instance: user => john password => 123

will be parsed to the following TCL list:

{{user john} {password 123}}

Also added the following tests:

* Redirection still works with RESP3

* Able to use a RESP3 client as a redirection client

* No duplicate invalidation messages when turning BCAST mode on after normal tracking

* Server is able to evacuate enough keys when num of keys surpasses limit by more than defined initial effort

* Different clients using different protocols can track the same key

* OPTOUT tests

* OPTIN tests

* Clients can redirect to the same connection

* tracking-redir-broken test

* HELLO 3 checks

* Invalidation messages still work when using RESP3, with and without redirection

* Switching to RESP3 doesn't disturb previous tracked keys

* Tracking info is correct

* Flushall and flushdb produce invalidation messages

**These tests achieve 100% line coverage for tracking.c using lcov.**

oranagra commented on tests/support/redis.tcl:
```diff
@@ -210,18 +210,40 @@ proc ::redis::redis_multi_bulk_read {id fd} {
 proc ::redis::redis_read_reply {id fd} {
     set type [read $fd 1]
     switch -exact -- $type {
+        _ -
```
Not sure if that's right.
Will return an empty line in case of NULL.
On the other hand, i'm not sure what's the right thing to use instead.

p.s. 
would be a good idea to add coverage for the rest of the RESP3 types in a followup PR.

nitaicaro replied:
You're right, I think we should return {} which is the return value for the RESP2 null representation:

```TCL
proc ::redis::redis_multi_bulk_read {id fd} {
    set count [redis_read_line $fd]
    if {$count == -1} return {}
```

oranagra commented on tests/unit/tracking.tcl:
```diff
@@ -1,116 +1,335 @@
-        assert {$keys eq {a:1 a:2 b:1 b:2}}
+        set keys1 [lsort [lindex [$rd_redirection read] 2]]
+        set keys2 [lsort [lindex [$rd_redirection read] 2]]
+        set res [lsort [list {*}$keys1 {*}$keys2]]
```
Any specific reason why you did all that renaming? (makes it harder to review, and also later track in blame log)

nitaicaro replied:
For the keys - no reason other than consistency and readability. For the clients - it was getting hard to keep track of which client was being used for what without more meaningful names. Perhaps revert key names and keep the clients?

oranagra commented on tests/unit/tracking.tcl:
```diff
@@ -1,116 +1,335 @@
+        # Preparing some message to consume on $rd_redirection so we don't get blocked
         r CLIENT TRACKING off
-        r CLIENT TRACKING on REDIRECT $redir NOLOOP
+        r CLIENT TRACKING on REDIRECT $redir_id
```
Why not let the second part of the test keep the same setup that was used for the first part? 
i.e. You show that one key isn't notified and the other is..
i think turning off the tracking and re-configuring it make the test weaker.

nitaicaro replied:
I agree

oranagra commented on tests/unit/tracking.tcl:
```diff
@@ -122,13 +341,59 @@ start_server {tags {"tracking"}} {
+        r config set tracking-table-max-keys $TRACKIING_TABLE_MAX_KEYS
+        # Eviction is done asynchronously so we want to give the server enough
+        # time to call the eviction function multiple times and increase effort.
+        after 1000
```
I don't like sleeps. maybe we can use a `wait_for_condition` instead?
If not, we at least need to test it in github actions with valgrind to see that timing is ok.

nitaicaro replied:
I'll replace it with a no sleep solution

itamarhaber reviewed (approved):
Wow, thanks for this. My TCL isn't what it used to be, so this LGTM :)

oranagra commented on tests/unit/tracking.tcl:
```diff
@@ -1,116 +1,335 @@
+        r CLIENT TRACKING on REDIRECT $redir_id
+        r SET key1 1
+        r GET key1
+        r INCR key1
+        r INCR key2 ; # This key should not be notified, since it wasn't fetched.
```
@nitaicaro didn't you recently change the order of these lines in #7871
using this order you can't make sure the incr for key2 didn't send an invalidation (without blocking the tests)

nitaicaro replied:
Whoops! My bad

oranagra commented on tests/unit/tracking.tcl:
```diff
@@ -1,116 +1,335 @@
-        set keys [lsort [lindex [$rd1 read] 2]]
-        assert {$keys eq {a b c}}
+        clean_all
+        r CLIENT TRACKING on BCAST REDIRECT $redir_id
```
Instead of building on the previous test, you start over. if you keep that, the prev test can be deleted, and this one add an assertion to verify the results of the CLIENT TRACKING command. (although maybe unnecessary since any error response will throw an exception)

many of the redis tests are incrementally building one on top of the other (i.e. each one continues from the state of the one before it), i'm not sure why you took that `clean_all` approach, if we keep it, some tests can be deleted or integrated into the tests that follow them, alternatively maybe you should consider keeping the old approach for many of the tests, and use clean_all only when necessary.

one difference between these approaches, is that with the old approach, was soon as one test fails, the rest will fail too. and with the one one each test is independent.
but i'm not sure that matters much.

oranagra reviewed (commented):
WOW. thank you @nitaicaro.
i skimmed over the changes and added a few random comments.

maybe later, someone need to have another careful review at the tests to check they're not missing anything.
i.e. i understand the C code is now 100% covered, that's great, but maybe some validations in the tests side can be improved to verify correct behavior  (i.e like the issue #7871 fixed).

oranagra commented:
@nitaicaro you did a force push containing a rebase, and some changes of your own?
please avoid that, i have no way to see what you changed.
it's fine to edit your code and force push (github will let me diff the commits), but if you also rebased in the same push, the changes are being mixed.
i suppose in this case the rebase didn't bring any changes to tracking.tcl, and redis.tcl, so all the changes there are yours.

i skimmed over the changes, i see you reduced the use of `clean_all` to just 2 places, the rest of the (both old and new) test just incrementally build on top of each other.
and i see you reverted the key name changes.
anything else you wanna point out for an easier review?

please temporarily modify `.github/workflows/daily.yml` to run on your PR (possibly adding `--single unit/tracking` to all the `./runtest` lines, so we can see how it works on all the slower platforms and valgrind.

p.s. it seems you have a conflict, i thought the rebase was in order to solve it. not sure what went wrong.

nitaicaro commented:
Sorry about that @oranagra and noted. 

I also changed the way the new NULL type is parsed. You can see the difference in redis.tcl. I also fixed a test that was acting a bit funny (`RESP3 Client gets tracking-redir-broken push message after cached key changed when rediretion client is terminated`) and removed the sleep from the `Server is able to evacuate enough keys when num of keys surpasses limit by more than defined initial effort` test.

oranagra commented:
@nitaicaro 
you'll need to remove this part from the daily yaml:
```
branches:
      # any PR to a release branch.	      # any PR to a release branch.
      - '[0-9].[0-9]'
```

and if you intend to test it on your own repo (pushing to another branch rather than the one used for this PR. i.e. if you're gonna run a lot of tests and don't wanna bother us seeing them), you'll need to remove this:
```
if: github.repository == 'redis/redis'
```

the reason i'm worried is that with deferred clients, you can't assume that by the time tcl reached the next line, redis actually processed the command you sent to it.
you may need something like this to make "sure" it was executed.
```
$rd flush
after 100
```
or better yet, some `wait_for_condition` loop looking at the flags of the CLIENT list.
possibly adding a CLIENT INFO (like the one discussed in #8000 and flags that discussed in #7995)

madolson commented:
+1 about the thankyou for adding tests. It's boring work people usually don't like doing 👍 

The one function I see that is "covered" but not validated is flushall/flushdb send invalidation messages.

oranagra commented:
it looks like there are no test timing issues that jump out. maybe we'll just have to deal with them when they come.

@nitaicaro can you please validate the correct behavior of flushdb that @madolson pointed out.
maybe @eliblight can review the tests and come up with other things that should / could be validated.

madolson commented:
@nitaicaro and @eliblight work together btw. He's already reviewed it.

eliblight commented:
indeed i have

On Wed, 4 Nov 2020, 23:03 Madelyn Olson, <notifications@github.com> wrote:

> @nitaicaro <https://github.com/nitaicaro> and @eliblight
> <https://github.com/eliblight> work together btw. He's already reviewed
> it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redis/redis/pull/7998#issuecomment-721973587>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABYF35VXAJROCVY6ETBCBBDSOG6RPANCNFSM4TET5HNQ>
> .
>

nitaicaro commented:
Implemented your suggestions @oranagra.

oranagra commented:
@nitaicaro tests are passing (at least 2 more times), i guess we're good to merge this.
please revert the changes to the daily CI, and see if the top comment of the PR needs any update.
then we can squash-merge it (using the top comment as the commit comment)
thank you.

nitaicaro commented:
@oranagra done

madolson commented:
I did glance through the difference from the last review, and didn't see anything wrong.
